Data_Engine/Query/Linq/FilterBy.cs

```diff
@@ -0,0 +1,162 @@
+        [Input("objects", "List of objects to be filtered. All objects in the list should be of similar type")]
+        [Input("propertyName", "The name of the property to filter the objects by")]
+        [Input("value", "The value of the selected property to filter the objects by")]
+        [Input("ignoreStringCase", "Ignore upper/lower case letters if the property filtered by is a text/string.")]
+        [Input("exactStringMatch", "If true checks if the property filtered by contains the text value (filtering must be a string/text)")]
+        [Output("filteredObjects", "The collection of objects that match the given filter conditions")]
+        public static List<T> FilterBy<T>(this List<T> objects, string propertyName = null, object value = null, bool ignoreStringCase = false, bool exactStringMatch = true)
+        {
+            if (objects == null || objects.Count == 0)
+            {
+                BH.Engine.Reflection.Compute.RecordWarning("No objects submitted to filter");
+                return new List<T>();
+            }
+
+            if (propertyName == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("propertyName cannot be null in order to filter the objects");
+                return new List<T>();
+            }
+
+            if (value == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("value cannot be null to filter the objects");
+                return new List<T>();
+            }
+
+            object firstObject = objects.Where(x => x != null).FirstOrDefault();
+            if(firstObject == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("All objects in the list to filter are null, please try with valid objects");
+                return new List<T>();
+            }
+
+            System.Type type = PropertyType(firstObject, propertyName);
+
+            List<T> filteredObjects = new List<T>();
+
+            if (type == typeof(System.String))
+            {
+                if (ignoreStringCase)
+                    filteredObjects.AddRange(objects.Where(x => x.PropertyValue(propertyName).ToString().Equals(value.ToString(), System.StringComparison.CurrentCultureIgnoreCase)).ToList());
+                else
+                    filteredObjects.AddRange(objects.Where(x => x.PropertyValue(propertyName).ToString().Equals(value.ToString())).ToList());
+
+                if (!exactStringMatch)
+                    filteredObjects.AddRange(objects.Where(x => x.PropertyValue(propertyName).ToString().Contains(value.ToString())).ToList());
+            }
+            else
+            {
+                dynamic val = System.Convert.ChangeType(value, type);
+                filteredObjects = objects.Where(x => (System.Convert.ChangeType(x.PropertyValue(propertyName), type) as dynamic) == val).ToList();
+            }
+
+            return filteredObjects.Distinct().ToList();
```

Comment:
Ha I see. Well I guess I should have read it all before commenting then 😋 .

I get that you want to have the exact matches first in your list but I wonder if that is really a feature you want. The user doesn't really have a way to know where in the list objects stop being exact matches so having them first is only partially helpful. On the other hand, it messes up the order of the input objects (say they where order alphabetically or something) which might be annoying for the user.

Reply:
I've refactored the statements slightly to try and avoid adding entries twice so this might be resolved by that. It might not be also, but I'm personally not too concerned about the order changing as we're adding the `OrderBy` as well/lists can be reordered, and nothing in our filter description suggests we will preserve current order - but maybe we can make that more explicit in the description if needed?